starlette/datastructures.py

```diff
@@ -290,6 +291,95 @@ def __repr__(self) -> str:
         return f"{self.__class__.__name__}({repr(items)})"
 
 
+class MultiDict(ImmutableMultiDict):
+    def __setitem__(self, key: typing.Any, value: typing.Any) -> None:
+        self.setlist(key, [value])
+        self._dict[key] = value
+
+    def __delitem__(self, key: typing.Any) -> None:
+        self._list = list(filter(lambda item: item[0] != key, self._list))
+        del self._dict[key]
+
+    def pop(self, key: typing.Any) -> typing.Any:
+        list_ = list(filter(lambda item: item[0] == key, self._list))
+        poped = self._dict.pop(key)
+        self.setlist(key, [v for _, v in list_[:-1]])
+        return poped
+
+    def popitem(self, key: typing.Any) -> typing.Tuple:
+        return key, self.pop(key)
+
+    def poplist(self, key: typing.Any) -> typing.List:
+        poped_list = list(filter(lambda item: item[0] == key, self._list))
+        self._list = list(filter(lambda item: item[0] != key, self._list))
+        self._dict.pop(key)
+        return [v for _, v in poped_list]
+
+    def clear(self) -> None:
+        self._dict.clear()
+        self._list.clear()
+
+    def setdefault(self, key: typing.Any, default: typing.Any = None) -> typing.Any:
+        if key not in self:
+            self._dict[key] = default
+            self._list.append((key, default))
+
+        return self[key]
+
+    def setlist(self, key: typing.Any, values: typing.List) -> None:
+        filter_obj = filter(lambda item: item[0] != key, self._list)
+        if not values:
+            values.append(None)
+        self._list = [*filter_obj, *((key, value) for value in values)]
+        self._dict[key] = self._list[-1][1]
+        # Case list is empty
+
+    def setlistdefault(
+        self, key: typing.Any, default_list: typing.List = None
+    ) -> typing.List:
+        if key not in self:
+            if default_list is None:
+                default_list = [None]  # or just []
+
+            self._list.extend(((key, value) for value in default_list))
+            self._dict[key] = default_list[-1]
+
+        return self.getlist(key)
+
+    def appendlist(self, key: typing.Any, value: typing.Any) -> None:
+        self._list.append((key, value))
+        self._dict[key] = value
+
+    def update(
+        self,
+        values: typing.Union[
+            "MultiDict",
+            typing.Mapping,
+            typing.List[typing.Tuple[typing.Any, typing.Any]],
+        ] = None,
+        **kwargs: typing.Any,
+    ) -> None:
```

Comment:
I think `update` should just overwrite any existing keys, rather than appending to them.

Reply:
Sorry, I didn't get it, could you provide some example

Reply:
Your test case gives:

```python
q = MultiDict([("a", "123"), ("b", "456")])
q.update({"a": "789"})
assert q.getlist("a") == ["123", "789"]
```

But I'd like to see:

```python
q = MultiDict([("a", "123"), ("b", "456")])
q.update({"a": "789"})
assert q.getlist("a") == ["789"]
q == MultiDict([("a", "789"), ("b", "456")])
```

Reply:
Thanks, and if multiple values were provided, we do smt like :

```python
q = MultiDict([("a", "123"), ("a", "456") ("b", "789")])
q.update([("a", "ABC"), ("a", "DEF")])
assert q.getlist("a") == ["ABC", "DEF"] # ?
```

Reply:
Yup 👍

Reply:
I guess the implementation will need to do something like:

```
items = ...
mapping = dict(items)
self._list = list(filter(lambda item: item[0] not in mapping, self._list)) + items
self._dict.update(mapping)
```

Reply:
close to what I did
```python
keys = [k for k, _ in itertools.chain(items_, kwargs.items())]
        self._list = [*((k, v) for k, v in self._list if k not in keys), *items_, *kwargs.items()]
        self._dict.update(self._list)
```